File: Data Structures/Find all symmetric pairs in an array of pairs.py
# Function to find all pairs that are a mirror of each other
def findPairs(pairs):
 
    # create an empty set of strings
    s = set()
 
    # do for each pair
    for (x, y) in pairs:
 
        # insert the current pair `(x, y)` into the set
        s.add((x, y))
 
        # if mirror pair `(y, x)` is seen before, print the pairs
        if (y, x) in s:
            print((x, y), "|", (y, x))
 
 
if __name__ == '__main__':
 
    pairs = [(3, 4), (1, 2), (5, 2), (7, 10), (4, 3), (2, 5)]
    findPairs(pairs)

File: Data Structures/Find all symmetric pairs in an array of pairs.cpp
#include <iostream>
#include <unordered_set>
using namespace std;
 
// Data structure to store a pair (or use `std::pair`)
struct Pair {
    int x, y;
};
 
// Function to find all pairs that are a mirror of each other.
// Here, we are creating a template that deduces the array size
// from its declared type
template<typename T, size_t n>
void findPairs(T const(& pair)[n])
{
    // create an empty set of strings
    unordered_set<string> set;
 
    // do for each pair
    for (int i = 0; i < n; i++)
    {
        // construct a pair `(x, y)` from `pair[i]`
        string p = "{" + to_string(pair[i].x) + ", " +
                        to_string(pair[i].y) + "}";
 
        // insert the current pair into the set
        set.insert(p);
 
        // construct a mirror pair `(y, x)` of `pair[i]`
        string reverse_pair = "{" + to_string(pair[i].y) + ", " +
                                    to_string(pair[i].x) + "}";
 
        // if the mirror pair is seen before, print the pairs
        if (set.find(reverse_pair) != set.end()) {
            cout << p << " | " << reverse_pair << endl;
        }
    }
}
 
int main()
{
    Pair pair[] = { {3, 4}, {1, 2}, {5, 2}, {7, 10}, {4, 3}, {2, 5} };
    findPairs(pair);
 
    return 0;
}